### ISSUE TITLE
Allow for TaskReference Interpolation within Pipeline

### ISSUE THREAD

**Issue by vtereso** [tektoncd/pipeline#1367]
# Expected Behavior
Without `PipelineResource`s ([google doc](https://docs.google.com/document/d/1u6qO7CPtDnTOZMYFQ5ARysSOy8lfFeVw83C_5BxAKsw/edit#heading=h.dqk738hxkkh)) `Pipeline` duplication happens as a consequence because there is no way to interchange the injection steps. In order to alleviate this shortcoming, taskRef injection will enable arbitrary `Task` to be similarly injected into the Pipeline.
# Actual Behavior
`Pipeline`s do not currently support this.

**Comment by bobcatfish** [comment]
> taskRef injection will enable arbitrary Task to be similarly injected into the Pipeline.

@vtereso can you elaborate a bit more on what taskRef injection would look like, maybe with an example?

**Comment by vtereso** [comment]
```
apiVersion: tekton.dev/v1alpha1
kind: Task
metadata:
  name: echo1
spec:
  steps:
  - name: comment
    image: docker.io/bash
    args: ["-c", 'echo 1']
---
apiVersion: tekton.dev/v1alpha1
kind: Task
metadata:
  name: echo1
spec:
  steps:
  - name: comment
    image: docker.io/bash
    args: ["-c", 'echo 2']
---
apiVersion: tekton.dev/v1alpha1
kind: Pipeline
metadata:
  name: echo-pipeline
spec:
  params:
  - name: taskRef
    type: string
  tasks:
  - name: echo
    taskRef: 
      name: $(params.taskRef)
---
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: echo-runner
spec:
  params:
  - name: taskRef
    value: echo1
  pipelineRef:
    name: echo-pipeline
```

**Comment by vtereso** [comment]
With this change, we would likely have strong enforcement where the taskRef must be a valid Kubernetes name after interpolation (the task must exist). Alternatively, an empty taskRef could be ignored when creating the `TaskRun` to get definition with optional injection points. In either situation, what this allows for in injection anywhere within the definition where `PipelineResources` only allow for prepending/postpending steps.

**Comment by bobcatfish** [comment]
Ah interesting, so the name of the Task could be interpolated! I think I could get behind that but I'm not quite sure about how well it addresses the use case you described:

> Without PipelineResources (google doc) Pipeline duplication happens as a consequence because there is no way to interchange the injection steps

The way that PipelineResources work right now, the logic being used to fetch your data is pretty fixed - if you use a git PipelineResource, you always get the same containers injected. I guess that storage is a bit different, b/c you can use GCS or GCS Build (or soon VolumeResource). But anyway I'm not sure you need the Task to be dynamic to get parity with how PipelineResources work today, it seems to me like having a git fetching `Task` and using that in the `Pipeline` gives you the same amount of potential duplication as using a git `PipelineResource` in a `Pipeline`?

**Comment by vtereso** [comment]
This creates a similar contract in terms of functionality, but rather than being at the `Task` level, it happens at the `Pipeline` level. For example, a `Task` can be interpolated to support both `mercurial` and `git` repository types, where the rest of the `Pipeline` would be identical. Without interpolated `Tasks`, this would create two distinct `Pipeline`. The combinatorics of not being able to  interchange these `Tasks` is what creates duplication.

**Comment by vtereso** [comment]
Another example, with the classical Clone->Build->Push `PIpeline`, the build `Task` will be different per language. Instead of having to maintain many different `Pipeline`, interpolating each language build `Task` consolidates this. This compliments `Pipelines` whether they use `PIpelineResources` or not.

**Comment by sbwsg** [comment]
We might want a slightly more complicated example in order to interrogate this idea a bit more.  Here's a mockup of what a git/hg/svn checkout pipeline might look like: 

```yaml
apiVersion: tekton.dev/v1alpha1
kind: Pipeline
metadata:
  name: checkout-the-thing
spec:
  params:
    - name: checkouterTaskRef
      type: string
      description: Task to checkout the thing
    - name: dir
      type: string
      description: Directory to checkout in to
      default: /some/where/or/other
    - name: repo
      type: string
      description: Repository to checkout the thing from
  tasks:
    - name: checkout-time-yeah
      taskRef:
        name: "$(params.checkouterTaskRef)"
      params:
        - name: outputDirectory
          value: "$(params.dir)"
        - name: repoURL
          value: "$(params.repo)"
```

And the PipelineRun to execute it:

```yaml
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: checkout-thing-with-git
spec:
  params:
  - name: checkouterTaskRef
    value: gitCheckout
  - name: repo
    value: "git@gitpub.dev:/myorg/myrepo.git"
  - name: dir
    value: "/src"
  pipelineRef:
    name: checkout-the-thing
```

And I guess the `gitCheckout` task (or any other VCS task) needs to follow the right shape of params as @vtereso already mentioned:

```yaml
apiVersion: tekton.dev/v1alpha1
kind: Task
metadata:
  name: gitCheckout
spec:
  inputs:
    params:
      - name: repoURL
        type: string
        description: The repo to check out
      - name: outputDirectory
        type: string
        description: The directory to checkout into
        default: "/workspace/gitCheckout"
  steps:
    - container: "alpine/git"
      args: ["$(params.repoURL)", "$(params.outputDirectory)"]
```

This feels like it's again a question of early vs late validation.  With a parameterised taskRef it's essentially impossible to perform any kind of early check that the task exists, is shaped how we expect, that the user has typed the parameter names with the right spelling, yada yada yada

**Comment by sbwsg** [comment]
Oh I should've included another task in the pipeline and a volume mount to demonstrate sharing between tasks.

**Comment by sbwsg** [comment]
I've updated the git clone example above to a mostly-working example (spent a couple hours getting it working locally minus the taskRef interpolation to make sure that the volume stuff was right).

The example clones a repo and `cat`s the README file.  It took me a long time to get everything lined up right with the volume configs.

Also, fun note, you can't name a volume "workspace" right now because we silently mount a workspace volume in the controller.  Spent too long figuring that out.

**Comment by bobcatfish** [comment]
> Without interpolated Tasks, this would create two distinct Pipeline. The combinatorics of not being able to interchange these Tasks is what creates duplication. The comparison to PipelineResources stems from their future plans (as I understand them) to support a sort of interface/inheritance model where an input something like a VCSResource that would have concrete handlers for git and the like as aforementioned.

Ah okay, I see what you mean. I do think that once we have better PipelineResources - and #1285 that we'll be able to meet your use case nicely tho. Maybe we can revisit this issue once #1285 is done or at least POC?

**Comment by bobcatfish** [comment]
Now we have workspaces which allow you to separate out how you get the files from what you do to the files.

I'm wondering if we can close this issue, reopen if it is requested again?
